frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -349,6 +351,39 @@ impl<T: Config> SignedSubmissions<T> {
 	}
 }
 
+/// Type that can be used to calculate the deposit base for signed submissions.
+///
+/// The deposit base is calculated as a geometric progression based on the number of signed
+/// submissions in the queue. The size of the queue represents the progression term.
+pub struct GeometricDepositBase<T> {
+	_marker: PhantomData<T>,
+}
+
+impl<T: Config> DepositCalculator<BalanceOf<T>> for GeometricDepositBase<T> {
+	/// The fixed deposit base upon which the geometric series is calculated.
```

Comment:
I don't see a need for these to be associated types; can you expand why? you can use `T::SignedDepositBaseIncreaseFactor` and so on directly.

These had to be associated types iff the input to these pallet was only a singular `type Deposit: DepositCalculator<Balance>`. Then all aspects of the deposit, including weight, len, and queue length need to be abstracted away and compacted into a single `T::Deposit::calculate(queue_len, solution_weight, solution_len)`. 

In either case, the existing `type BaseDeposit` and `type IncreaseFactor` don't serve any purpose.

Reply:
An alternative design that seems clean to me is to have: 

```
trait Config {
  type DepositLen: Convert<usize, Balance>;
  type DepositWeight: Convert<weight, Balance>;
  type DepositByte: Convert<u32, Balance>;
}
```

each in charge of calculating one dimension of the deposit. The final deposit would be the sum of two. The previous suggestion was merely hiding all 3 behind a single function and a custom trait instead of `Convert`. Albeit, even for the previous one, you could do `Convert<(usize, u32, Weight), Balance>`, but that's a bit too ugly.

Reply:
> I don't see a need for these to be associated types; can you expand why? you can use T::SignedDepositBaseIncreaseFactor and so on directly.

The goal was to keep the `trait DepositCalculator` as general as possible and to provide the implementors a way to fetch the `IncreaseFactor` and the `BaseDeposit` externally. I agree that these types are redundant when the implementor has access to `T` though, as it is in this implementation.

I will rethink this trait.

Reply:
So I think a good compromise which is clean IMO is to keep the `SIgnedDepositByte` and `SignedDepositWeight` as is (i.e. no refactor to use `Convert<A, Balance>`) and keep the current implementation of `fn deposit_for` returning the final deposit balance.

However, the I have removed the previous `trait DepositCalculator ` (which I agree was a bit convoluted) and added a new associated type to EPM's config that converts a usize to Balance. And implement a geometric progression on a type that implements convert.